src/layers/attention/self_attention.rs

```diff
@@ -390,6 +426,62 @@ impl Attention for SelfAttention {
     }
 }
 
+impl SelfAttention {
+    /// Compute query, key, and value with separate projections.
+    fn query_key_value_separate(
+        &self,
+        value: &Linear,
+        query: &Linear,
+        key: &Linear,
+        input: &Tensor,
+    ) -> Result<(Tensor, Tensor, Tensor), BoxedError> {
+        let query = query
+            .forward(input)
+            .context(QkvSnafu)?
+            .split_heads(self.attention_heads.n_query_heads)?;
+        let key = key
+            .forward(input)
+            .context(QkvSnafu)?
+            .split_heads(self.attention_heads.n_key_value_heads)?;
+        let value = value
+            .forward(input)
+            .context(QkvSnafu)?
+            .split_heads(self.attention_heads.n_key_value_heads)?;
+        Ok((query, key, value))
+    }
+
+    /// Compute query, key, and value with a single projection.
+    ///
+    /// Split heads before query, key, and value.
+    fn query_key_value_split_before(
+        &self,
+        qkv: &Linear,
+        input: &Tensor,
+    ) -> Result<(Tensor, Tensor, Tensor), BoxedError> {
+        let proj = qkv
+            .forward(input)
+            .context(QkvSnafu)?
+            .split_heads(self.attention_heads.n_query_heads)?;
+
+        let (_, _, _, all_heads_size) = proj.shape().dims4().context(QkvSnafu)?;
+        let head_size = all_heads_size / 3;
+
+        // Similar to chunk, but avoid intermediate Vec. Needs to be contiguous or
+        // matrix multiplication fails later.
+        let query = proj.narrow(3, 0, head_size).context(QkvChunkSnafu)?;
+        let key = proj
+            .narrow(3, head_size, head_size)
+            .context(QkvChunkSnafu)?;
```

Comment:
Don't we need to ensure contiguity here as well (and for `query`)?

Reply:
So far it doesn't seem to be necessary. Perhaps that's because of the transpose later for `key` in attention and for both because the only model that I tested uses rotary embeddings. We should probably add it on-need basis, because we do have quite a lot of operations that make strides non-contiguous and we probably don't want to make it contiguous in every case.

I was also kind of expecting that `matmul` would do this in case it's necessary.